### DISCUSSION TITLE
Substitute for --unsafe-perm in NPM 7 ?

### DISCUSSION THREAD

**Discussion by ingun37** [npm/feedback#121] [CLI] [answered]
I found `--unsafe-perm` became obselete in NPM 7.
https://github.com/npm/cli/pull/2196

What should I use instead?

**Comment by MylesBorins** [comment, marked as answer]
Can you expand on your use case a bit more? We no longer drop permissions when running shell scripts, we use the permissions of the folder.

**Reply by ingun37** [reply]
I get permission error when I install *puppeteer* globally.
```

npm ERR! code 1
npm ERR! path /usr/local/lib/node_modules/puppeteer
npm ERR! command failed
npm ERR! command sh -c node install.js

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-12-16T06_45_17_738Z-debug.log
```
And the `debug.log` outputs
```
...
330 error code 1
331 error path /usr/local/lib/node_modules/puppeteer
332 error command failed
333 error command sh -c node install.js
334 verbose exit 1
```
You can recreate the issue by following these steps on newly created ec2 ubuntu 20 instance ... but I think you can do it on any other linux environment.
```
sudo snap install node --classic --channel=15
sudo npm i -g puppeteer
```

I was asking because I thought I had the same problem on npm 6 and solved by using `--unsafe-perm`.

You said it uses permission of the folder. Which folder are you referring?

**Reply by MylesBorins** [reply]
@ingun37 is this still an issue with the latest version of the npm CLI?

**Reply by ingun37** [reply]
I’ll look into it and let you know soon

**Reply by ingun37** [reply]
@MylesBorins 
I'm still having the same problem with 7.5.2
```
320 info run puppeteer@6.0.0 install node_modules/puppeteer node install.js
321 info run puppeteer@6.0.0 install { code: 1, signal: null }
322 timing reify:rollback:createSparse Completed in 182ms
323 timing reify:rollback:retireShallow Completed in 0ms
324 timing command:install Completed in 26861ms
325 verbose stack Error: command failed
325 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/index.js:64:27)
325 verbose stack     at ChildProcess.emit (node:events:378:20)
325 verbose stack     at maybeClose (node:internal/child_process:1067:16)
325 verbose stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
326 verbose pkgid puppeteer@6.0.0
327 verbose cwd /home/ubuntu
328 verbose Linux 5.4.0-1029-aws
329 verbose argv "/snap/node/3663/bin/node" "/usr/local/bin/npm" "i" "-g" "puppeteer"
330 verbose node v15.8.0
331 verbose npm  v7.5.2
332 error code 1
333 error path /usr/local/lib/node_modules/puppeteer
334 error command failed
335 error command sh -c node install.js
336 verbose exit 1
```

**Reply by MylesBorins** [reply]
@ingun37 as a next step here you should open an issue over at https://github.com/npm/cli/issues

**Reply by robross0606** [reply]
How is "you should open a ticket" considered the answer here?

**Reply by alexandernst** [reply]
ffs I'm sick and tired of all you js people.... can you please just stop making breaking changes every 2 months in all the tooling jesus christ... 

npm can be ran inside of a ephemeral container inside a ci/cd pipeline as part of a build process, which I couldn't, wouldn't and shouldn't bother to run with a user that isn't root (the default in a container) because I literally don't care at all about any security implications becase **it's running inside a container that will have a ~2min live span**.

please. I'm begging you. stop applying random breaking-changes shit.

**Comment by eybel** [comment]
It seems that unsafe-perm=true is gone??? Im trying to upload an app to AWS ElasticBeanstalk with Node 16 and this doesnt work now... any replacement? I cant find it in the docs!

**Reply by Jimmar** [reply]
I managed to get it to work by adding an `.npmrc` file to the root of the project and putting this in it
```
unsafe-perm=true
```

**Reply by tmurvv** [reply]
Does this turn enable unsafe-perm=true for the whole life-cycle of the app? I am looking for a solution that only enables it during build-time, not the whole life-cycle of the app.

**Reply by ljharb** [reply]
You can set an env var during the build if you like - NPM_CONFIG_UNSAFE_PERM, i think.

**Reply by PeterDaveHello** [reply]
So `unsafe-perm` wasn't removed entirely?
